core/src/main/java/com/sap/ai/sdk/core/DeploymentResolver.java

```diff
@@ -19,99 +19,104 @@
-  }
-
-  private Optional<String> getDeploymentIdByModel(
-      @Nonnull final String resourceGroup, @Nonnull final AiModel model) {
-    return cache.getOrDefault(resourceGroup, new HashSet<>()).stream()
-        .filter(deployment -> isDeploymentOfModel(model, deployment))
-        .findFirst()
-        .map(AiDeployment::getId);
+  String getDeploymentIdByModel(@Nonnull final String resourceGroup, @Nonnull final AiModel model)
+      throws DeploymentResolutionException {
+    final Predicate<AiDeployment> predicate = deployment -> isDeploymentOfModel(model, deployment);
+    return resolveDeployment(resourceGroup, predicate)
+        .orElseThrow(
+            () ->
+                new DeploymentResolutionException(
+                    "No running deployment found for model: " + model.name()));
   }
 
   /**
    * Get the deployment id from the scenario id. If there are multiple deployments of the * same
    * model, the first one is returned.
    *
-   * @param aiCoreDestination the configured connection to the AI Core service.
    * @param resourceGroup the resource group, usually "default".
    * @param scenarioId the scenario id, can be "orchestration".
    * @return the deployment id.
-   * @throws NoSuchElementException if no running deployment is found for the scenario.
+   * @throws DeploymentResolutionException if no running deployment is found for the scenario.
    */
   @Nonnull
   String getDeploymentIdByScenario(
-      @Nonnull final AiCoreDestination aiCoreDestination,
-      @Nonnull final String resourceGroup,
-      @Nonnull final String scenarioId)
-      throws NoSuchElementException {
-    return getDeploymentIdByScenario(resourceGroup, scenarioId)
-        .orElseGet(
-            () -> {
-              resetCache(aiCoreDestination, resourceGroup);
-              return getDeploymentIdByScenario(resourceGroup, scenarioId)
-                  .orElseThrow(
-                      () ->
-                          new NoSuchElementException(
-                              "No running deployment found for scenario: " + scenarioId));
-            });
+      @Nonnull final String resourceGroup, @Nonnull final String scenarioId)
+      throws DeploymentResolutionException {
+    final Predicate<AiDeployment> predicate =
+        deployment -> scenarioId.equals(deployment.getScenarioId());
+    return resolveDeployment(resourceGroup, predicate)
+        .orElseThrow(
+            () ->
+                new DeploymentResolutionException(
+                    "No running deployment found for scenario: " + scenarioId));
+  }
+
+  @Nonnull
+  Optional<String> resolveDeployment(
+      @Nonnull final String resourceGroup, @Nonnull final Predicate<AiDeployment> predicate) {
+    var deployment = getCachedDeployment(resourceGroup, predicate);
+    if (deployment.isPresent()) {
+      return deployment;
+    }
+    // double-checked locking to allow for safe cache resets under parallel access
+    synchronized (cache) {
```

Comment:
(Comment)

Not a fan of the double lock code. Can't this be solved with convenience API somewhere from the (vanilla) java collection API?

Reply:
Yeah, also don't like implementing this manually. But not sure. We can't directly use computeIfAbsent, because that would only compute when the resource group isn't found in the cache. I don't know of any other convenience / tooling around this that standardizes this..

Reply:
I see. Maybe one option to do this properly would be a real cache implementation, that has invalidation handlers - which would trigger when _predicate_ returned _empty_. But I understand that sounds like overkill for the current code line.